### PR TITLE
improvement: Extract shared mtags without scalameta dependency

### PR DESCRIPTION
**Pull request by rochala**
This PR is a part of ongoing work to make the presentation compiler more stable by moving its implementation to the dotty repository. All future releases of PC implementation, will from that point be released with the compiler. For further explanation , read [this](https://contributors.scala-lang.org/t/stable-presentation-compiler-api/6139).

This PR is big and focuses on multiple aspects.
Firstly, it splits shared code between mtags2 and mtags3 to a new module. It is called `mtags-shared` and can't have any Scala dependencies (most importantly, scalameta parser).

Because of that change, 2 mtags capabilities had to be reimplemented:
- KeywordCompletions - they are not yet ported, but they will be when mtags are moved to dotty,
- InlineValueProvider - logic for scala3 was reimplemented to use untyped trees instead of scalameta parser.

Other than that, I also created logic in `build.sbt` to allow testing metals with presentation compiler published directly from dotty. To enable it, the version from which the release cycle will begin, needs to be set:
```scala
def isStablePC(v: Option[(Long, Long)]): Boolean =
  false // replace after dotty starts publishing stable PC
// v.contains((3, 3)) // to be changed according to the first version with Stable PC
```

When mtags are finally moved to dotty, all changes to mtags should be made in dotty with optional porting them here to support older versions of Scala3 (this won't be a priority as Scala3 LTS will be the first version providing its PC, and everyone is expected to use it instead of older versions).

I want this PR to be merged before moving mtags to dotty, as it changes directory structure and conflict resolution is way harder.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -243,6 +259,28 @@ lazy val interfaces = project
...
+  .in(file("mtags-shared"))
+  .settings(sharedSettings)
+  .settings(
+    moduleName := "mtags-shared",
+    crossTarget := target.value / s"scala-${scalaVersion.value}",
+    crossScalaVersions := {
+      V.supportedScalaVersions ++ V.nightlyScala3Versions
+    },
+    crossVersion := CrossVersion.binary,
+    Compile / packageSrc / publishArtifact := true,
```
Comment (tgodzik) [inline build.sbt:272]: That's the default in the build, can be removed

File: build.sbt [hunk 2]
```diff
@@ -266,12 +304,26 @@ val mtagsSettings = List(
...
+      (ThisBuild / baseDirectory).value / "mtags",
+      scalaVersion.value,
+    ),
   ),
   // @note needed to deal with issues with dottyDoc
   Compile / doc / sources := Seq.empty,
+  libraryDependencies ++= crossScalaStablePC(
+    scalaVersion.value,
+    ifStablePC = List(
+      "org.scala-lang" %% "scala3-presentation-compiler" % scalaVersion.value
```
Comment (tgodzik) [inline build.sbt:323]: Are we sure this will be the artifact? Maybe we could add these changes later?

File: tests/unit/src/test/scala/tests/codeactions/ExtractValueLspSuite.scala [hunk 3]
```diff
@@ -144,15 +144,15 @@ class ExtractValueLspSuite
...
-       |    println("Hello!"); 
-       |    val newValue = 1 + 2
-       |    method2(1 + method2(newValue))
-       |  }
-       |}
-       |""".stripMargin,
+    s"""|object Main {
+        |  def method2(i: Int) : Int = ???
+        |  def method1(s: String): Unit = {
+        |    println("Hello!");${' '}
```
Comment (tgodzik) [inline tests/unit/src/test/scala/tests/codeactions/ExtractValueLspSuite.scala:150]: Why is this change needed? Do we even need to change this file?

File: mtags/src/main/scala-3/scala/meta/internal/pc/PcInlineValueProviderImpl.scala [hunk 4]
```diff
@@ -55,13 +53,44 @@ final class PcInlineValueProviderImpl(
...
         RangeOffset(defPos.start, defPos.end),
+        definitionRequiresBrackets(definition.tree.rhs)(using newctx),
         deleteDefinition,
       )
 
       (defEdit, refsEdits)
     end for
   end defAndRefs
 
+  // format: off
```
Comment (tgodzik) [inline mtags/src/main/scala-3/scala/meta/internal/pc/PcInlineValueProviderImpl.scala:64]: Please don't use `format: off` we can introduce settings if needed, but I would rather avoid it if possible.

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala [hunk 5]
```diff
@@ -528,6 +528,45 @@ class CompletionKeywordSuite extends BaseCompletionSuite {
        |""".stripMargin,
   )
 
+  check(
+    "extends-with-class".tag(IgnoreScala2),
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala:532]: This should also work in Scala 2. Is it not working because of your change or is it working in Scala 3 because of your change? In the latter case let's create an issue for Scala 2 here.

File: tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala [hunk 6]
```diff
@@ -273,14 +272,14 @@ class CompletionArgSuite extends BaseCompletionSuite {
...
-       |""".stripMargin,
+    s"""|$$$$foo =${' '}
+        |$$$$foo = $${1:???}, | = $${2:???}
+        |""".stripMargin,
     topLines = Option(2),
     compat = Map(
-      "3" -> """|$$foo = 
-                |$$foo = ${1:???}, $$bar = ${2:???}
-                |""".stripMargin
+      "3" -> s"""|$$$$foo =${' '}
```
Comment (tgodzik) [inline tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:280]: Is this needed? Why not leave `_` <- seems like something Intellij is forcing you to do :D

**Review by tgodzik — COMMENTED** [review]
Thanks for all the work, I left some initial comments. I have around 20 files more to go through.

File: tests/unit/src/test/scala/tests/codeactions/ExtractValueLspSuite.scala (hunk 3 above)
Comment (rochala) [reply tests/unit/src/test/scala/tests/codeactions/ExtractValueLspSuite.scala:150]: Trailing whitespaces are not the best things to have in your code as they break navigation. I spent some time on this test because my editor (not IntelliJ), removed it, and it turned out to be important in this test. In my opinion such whitespaces should be explicitly written as important. Maybe the execution of this could be better, lets say by using a space assigned to val `$whitespace`

File: tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala (hunk 5 above)
Comment (rochala) [reply tests/cross/src/test/scala/tests/pc/CompletionKeywordSuite.scala]: I forgot to remove it. It works for both Scala versions.

File: build.sbt (hunk 2 above)
Comment (rochala) [reply build.sbt]: I reverted those changes. I'll add them when PR for dotty is created and ready.

File: tests/unit/src/test/scala/tests/codeactions/ExtractValueLspSuite.scala [hunk 7]
```diff
@@ -144,15 +144,15 @@ class ExtractValueLspSuite
...
-    """|object Main {
-       |  def method2(i: Int) : Int = ???
-       |  def method1(s: String): Unit = {
-       |    println("Hello!"); 
-       |    val newValue = 1 + 2
-       |    method2(1 + method2(newValue))
-       |  }
-       |}
-       |""".stripMargin,
+    s"""|object Main {
```
Comment (tgodzik) [inline tests/unit/src/test/scala/tests/codeactions/ExtractValueLspSuite.scala:147]:
```suggestion
    """|object Main {
```
it's no longer necessary

File: tests/unit/src/test/scala/tests/codeactions/ExtractValueLspSuite.scala (hunk 3 above)
Comment (tgodzik) [reply tests/unit/src/test/scala/tests/codeactions/ExtractValueLspSuite.scala]: Sure, makes sense.

File: build.sbt (hunk 1 above)
Comment (rochala) [reply build.sbt:256]: This was overriden here:
https://github.com/rochala/metals/blob/ef7e82144bd0f6f3baf0052e1088de9e4861aa3e/build.sbt#L66

Comment (tgodzik) [reply build.sbt:256]: Wait, that would mean they are overriden to true on CI, where it will be released. Not locally, sure.

Comment (rochala) [reply build.sbt:256]: Yea, but when you are working with metals, it will be very annoying to add this line every time you want to deploy mtags-shared locally, just to test if something works in dotty.

**Comment by rochala** [comment]
Will do.
